clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -3946,6 +4080,32 @@ static Address emitArraySubscriptGEP(CodeGenFunction &CGF, Address addr,
   return Address(eltPtr, CGF.ConvertTypeForMem(eltType), eltAlign);
 }
 
+static bool GetFieldOffsetInBits(CodeGenFunction &CGF, const RecordDecl *RD,
+                                 const FieldDecl *FD, uint64_t &Offset) {
+  ASTContext &Ctx = CGF.getContext();
+  const ASTRecordLayout &Layout = Ctx.getASTRecordLayout(RD);
+  unsigned FieldNo = 0;
+
+  for (const Decl *D : RD->decls()) {
+    if (const auto *Record = dyn_cast<RecordDecl>(D))
+      if (GetFieldOffsetInBits(CGF, Record, FD, Offset)) {
+        Offset += Layout.getFieldOffset(FieldNo);
+        return true;
+      }
+
+    if (const auto *Field = dyn_cast<FieldDecl>(D))
+      if (FD == Field) {
+        Offset += Layout.getFieldOffset(FieldNo);
+        return true;
+      }
+
+    if (isa<FieldDecl>(D))
+      ++FieldNo;
+  }
```

Comment:
I am deeply confused by the code in this function.  Why are we visiting every member of the record?

It looks like the caller has the exact field path you need, but it's throwing that away and doing a brute-force recursive search.  Just take a `NamedDecl` here and require it to be either a `FieldDecl` or an `IndirectFieldDecl`.  Or better yet, add that as a method to `ASTRecordLayout`.

Reply:
If you call `getFieldIndex` for a `FieldDecl` that's not in a top-level of the `RecordDecl`, it will assert. So I need to go through each field one at a time, because the method is broken.

I was planning on adding that to the record layout contexts, but because this patch has been through the wringer, I didn't want to throw yet another reason for people to hate it.

Reply:
Let's start by breaking that out as a separate patch.  Just make a method on `ASTRecordLayout` that takes an `IndirectFieldDecl` and recursively adds up the offsets for every step in the path.

Reply:
Could I please just do that as a follow-up patch? I don't have a great understanding of `ASTRecordLayout` and don't want to get bogged down in yet another endless review cycle.